numpy/random/_generator.pyx

```diff
@@ -3021,9 +3021,37 @@ cdef class Generator:
         ...    print(i, "wells drilled, probability of one success =", probability)
 
         """
+
+        cdef bint is_scalar = True
+
+        p_arr = <np.ndarray>np.PyArray_FROM_OTF(p, np.NPY_DOUBLE, np.NPY_ALIGNED)
+        is_scalar = is_scalar and np.PyArray_NDIM(p_arr) == 0
+        n_arr = <np.ndarray>np.PyArray_FROM_OTF(n, np.NPY_INT64, np.NPY_ALIGNED)
+        is_scalar = is_scalar and np.PyArray_NDIM(n_arr) == 0
+
+        if not is_scalar:
+            check_array_constraint(n_arr, 'n', CONS_POSITIVE_NOT_NAN)
+            check_array_constraint(p_arr, 'p', CONS_BOUNDED_GT_0_1)
+            # Check that the choice of negative_binomial parameters won't result in a
+            # call to the poisson distribution function with a value of lam too large.
+            max_lam_arr = (1 - p_arr) / p_arr * (n_arr + 10 * np.sqrt(n_arr))
+            if np.any(np.greater(max_lam_arr, POISSON_LAM_MAX)):
+                raise ValueError("n too large or p too small")
+        else:
+            _dp = PyFloat_AsDouble(p)
```

Comment:
In the scalar path, think you can use something like
```
cdef double *_dp
cdef int64_t *_in

_in = <int64_t*>np.PyArray_DATA(n_arr)[0]
_dp = <double*>np.PyArray_DATA(p_arr)[0]
check_constraint(<double>_in[0], 'n', CONS_POSITIVE_NOT_NAN)
check_constraint(_dp[0], 'p', CONS_BOUNDED_GT_0_1)
```

since you already have paid the price for the conversion above to arrays, so can just grab the first element form each when scalar.

Reply:
Okay, this surfaced some issues with my implementation: while numpy's `binomial` accepts floating point values for the `n` argument, they're truncated to integer values. On the other hand, `negative_binomial` used to accept floating point arguments without truncation, so the `n` array should actually be of type double.

Reply:
Done now, hopefully!